# Linear Regression

In [1]:
# import packages 
import statsmodels.api as sm
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn import datasets 
from sklearn import metrics
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

In [2]:
# Variables in house data: 
# CRIM - per capita crime rate by town
# ZN - proportion of residential land zoned for lots over 25,000 sq.ft.
# INDUS - proportion of non-retail business acres per town.
# CHAS - Charles River dummy variable (1 if tract bounds river; 0 otherwise)
# NOX - nitric oxides concentration (parts per 10 million)
# RM - average number of rooms per dwelling
# AGE - proportion of owner-occupied units built prior to 1940
# DIS - weighted distances to five Boston employment centres
# RAD - index of accessibility to radial highways
# TAX - full-value property-tax rate per $10,000
# PTRATIO - pupil-teacher ratio by town
# LSTAT - % lower status of the population
# MEDV - Median value of owner-occupied homes in $1000's

house = pd.read_csv('house.csv') 


In [3]:
house.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

   LSTAT  MEDV  
0   4.98  24.0  
1   9.14  21.6  
2   4.03  34.7  
3   2.94  33.4  
4   5.33  36.2

In [4]:
house.shape 

(506, 13)

# Simple Linear Regression 

* Simple linear regression is an approach for predicting a quantitative response using a single feature (or "predictor" or "input variable")

* It takes the following form:
* $y=\beta_{0}+\beta_{1}x$
 


* $y$  is the target variable
* $x$  is the predictor variable
* $\beta_{0}$  is the intercept
* $\beta_{1}$  is the coefficient for $x$

* $\beta_{0}$  and  $\beta_{1}$  are called the model coefficients

To build the regression model, we first estimate the values of these coefficients using data for target and predictor variables. Once we learn these coefficients, we can use the model to make predictions.



# Scikit-Learn
- Used primarily for machine learning models where predications are emphasized
- Pros: 
    - Comprehensive machine learning models available 
- Cons: 
    - Not easy to implement statistical models (e.g. obtaining p-values in linear regression is not obvious)
    - Statistical models are less comprehensive than Statsmodels

In [5]:
X = house[['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'LSTAT']]
# X = house[['ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'LSTAT']]
# X = house[['CRIM', 'ZN', 'INDUS', 'CHAS']]
y = house['MEDV'] 


In [6]:
X

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     PTRATIO  LSTAT  
0       15.3   4.98  
1       17.8   9.14  
2       17.8   4.03  
3       18.7   2.94  
4       18.7   5.33  
..       ...    ...  
501     21.0   9.67  
502     21.0   9.08  
503     21.0   5.64  
504     21.0   6.48  
505     21.0   7.88  

[506 rows x 12 columns]

In [7]:
y

0      24.0
1      21.6
2      34.7
3      33.4
4      36.2
       ... 
501    22.4
502    20.6
503    23.9
504    22.0
505    11.9
Name: MEDV, Length: 506, dtype: float64

## Train Test Split

Split the data into a training set and a testing set. We will train out model on the training set and then use the test set to evaluate the model.

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)


In [9]:
house.shape, X_train.shape, X_test.shape

((506, 13), (354, 12), (152, 12))

In [10]:
model1 = LinearRegression(fit_intercept=True)

model1.fit(X_train, y_train)


LinearRegression()

## Model Evaluation

Evaluate the model by checking out it's coefficients and how we can interpret them.

In [11]:
model1.intercept_


43.65366471149055

In [12]:
model1.coef_ 

array([-9.26879138e-02,  5.04105955e-02,  1.82221503e-02,  3.91252801e+00,
       -1.80480339e+01,  3.12354295e+00,  1.39131909e-02, -1.40907664e+00,
        2.49535042e-01, -1.05362854e-02, -9.43034937e-01, -6.24669882e-01])

In [13]:
X.columns 

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'LSTAT'],
      dtype='object')

In [14]:
coeff_df = pd.DataFrame(model1.coef_, X.columns, columns=['Coefficient'])
coeff_df

Coefficient
CRIM       -0.092688
ZN          0.050411
INDUS       0.018222
CHAS        3.912528
NOX       -18.048034
RM          3.123543
AGE         0.013913
DIS        -1.409077
RAD         0.249535
TAX        -0.010536
PTRATIO    -0.943035
LSTAT      -0.624670

## Predictions

In [15]:
predictions1 = model1.predict(X_test)


In [16]:
len(predictions1), len(X_test)

(152, 152)

In [17]:
predictions1

array([40.03785327, 27.40224731, 17.85321818, 17.10505344, 31.19095264,
       32.06652237, 38.4114798 ,  8.7174117 , 33.64633058,  6.64540382,
       30.48745243, 13.16058298, 16.87683884, 17.37333593, 25.10886311,
       20.42443494,  8.09113741, 33.1950366 , 28.69866151, 24.59585122,
       11.70951566, 20.04096199, 22.60154678, 24.4107513 , 33.94097328,
       18.34739763, 32.59991135, 18.45636083, 27.45469724, 34.38897463,
       19.80846035, 18.44430624, 37.21804287, 45.0511227 , 30.03225646,
       21.90916655, 17.42423431, 18.25523309,  5.22707775, 31.08100189,
       24.15206201, 17.40703557, 34.06821615, 13.26918202, 17.30274583,
       25.27905454, 30.37122284, 15.83382862, 26.96377387, 22.97483753,
       32.10135623, 37.12760939, 23.35315653, 19.45862602, 30.25501291,
       -0.86908712, 20.37753875, 16.63692238, 23.17937712, 21.06519058,
       30.59174079,  2.58632755, 16.15022632, 20.32890045, 11.99843816,
       24.2284151 , 23.96660095, 19.82595465, 16.83793454, 19.48

## Regression Evaluation Metrics

In [18]:
print('MAE:', metrics.mean_absolute_error(y_test, predictions1))
print('MSE:', metrics.mean_squared_error(y_test, predictions1))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, predictions1))) # numpy func np.sqrt takes sq root


MAE: 3.9560561666860776
MSE: 29.697713321530717
RMSE: 5.4495608374923865


## K-Fold Cross-Validation

In [19]:
model2 = LinearRegression(fit_intercept=True)

model2.fit(X_train,y_train)

metrics = cross_val_score(model2, X, y, scoring='neg_mean_squared_error', cv=5)

print(f'MSE for k folds: \n {-metrics}')  

print(f'RMSE for k folds: \n {np.sqrt(-metrics)}')  

print(f'Average RMSE across k folds: \n {np.sqrt(-metrics).mean()}')  


MSE for k folds: 
 [12.5537259  26.48996378 34.40346886 77.20099629 30.98760414]
RMSE for k folds: 
 [3.54312375 5.14684017 5.86544703 8.78640975 5.56665107]
Average RMSE across k folds: 
 5.78169435468386


## Regularization

In [20]:
model3 = Lasso()

model3.fit(X, y) 


Lasso()

In [21]:
model3.intercept_

45.0210107765927

In [22]:
coeff_df = pd.DataFrame(model3.coef_,X.columns,columns=['Coefficient'])
coeff_df

Coefficient
CRIM       -0.074012
ZN          0.050018
INDUS      -0.000000
CHAS        0.000000
NOX        -0.000000
RM          0.828241
AGE         0.022704
DIS        -0.660557
RAD         0.250041
TAX        -0.015923
PTRATIO    -0.702196
LSTAT      -0.788264

In [23]:
model4 = Ridge()

model4.fit(X, y) 


Ridge()

In [24]:
model4.intercept_

36.73125433672166

In [25]:
coeff_df = pd.DataFrame(model4.coef_,X.columns,columns=['Coefficient'])
coeff_df

Coefficient
CRIM       -0.116365
ZN          0.047992
INDUS      -0.017575
CHAS        2.708749
NOX       -11.423121
RM          3.694807
AGE        -0.002670
DIS        -1.381758
RAD         0.270548
TAX        -0.013303
PTRATIO    -0.856052
LSTAT      -0.563387

# StatsModels
- Used primarily for statiscal models where explanations of what causes the y-variable to change is emphasized
- Pros: 
    - Easy to obtain coefficients, p-values, R2, etc (similar to R outputs) 
- Cons: 
    - Not easy to perform machine learning tasks (split data into training and test sets, get performance metrics such as precision, recall, accuracy...)
    - Does not have compreshensive machine learning models as in Scikit-Learn 

In [26]:
X1 = house["RM"] 
X1 = sm.add_constant(X1) 
y = house["MEDV"] 

model5 = sm.OLS(y, X1).fit() 

predictions5 = model5.predict(X1)

model5.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.484
Model:                            OLS   Adj. R-squared:                  0.483
Method:                 Least Squares   F-statistic:                     471.8
Date:                Sun, 24 Sep 2023   Prob (F-statistic):           2.49e-74
Time:                        10:32:53   Log-Likelihood:                -1673.1
No. Observations:                 506   AIC:                             3350.
Df Residuals:                     504   BIC:                             3359.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -34.6706      2.650    -13.084      0.000     -39.877     -29.465
RM             9.1021      0.419     21.722      0.000       8.279       9.925
==============================================================================
Omnibus:                      102.585   Durbin-Watson:                   0.684
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              612.449
Skew:                           0.726   Prob(JB):                    1.02e-133
Kurtosis:                       8.190   Cond. No.                         58.4
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Interpretation

Interpreting the RM coefficient ($\beta_{1}$)

A unit increase in RM (average number of rooms in the house) is associated with a $\$$9,100 increase in median house price in that town. 


## Prediction
Let's say that there was a new town with an average of 5 rooms in each house. What would we predict for the median house price?

$y=-34.6706+9.1021x$
 
$y=-34.6706+9.1021 \times 5 = \$10,840$


## Prediction using Statsmodels

In [27]:
X = pd.DataFrame({'const': [1], 'RM': [5]})[['const', 'RM']]

prediction5 = model5.predict(X)

print('A town with an average of {} rooms per house is expected to have a median house value of ${:.0f}.'.format(int(X.RM),float(prediction5)*1000))


A town with an average of 5 rooms per house is expected to have a median house value of $10840.


C:\Users\Test\AppData\Local\Temp\ipykernel_14692\543673117.py:5: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead
  print('A town with an average of {} rooms per house is expected to have a median house value of ${:.0f}.'.format(int(X.RM),float(prediction5)*1000))
C:\Users\Test\AppData\Local\Temp\ipykernel_14692\543673117.py:5: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  print('A town with an average of {} rooms per house is expected to have a median house value of ${:.0f}.'.format(int(X.RM),float(prediction5)*1000))


In [28]:
X

const  RM
0      1   5

# Multiple Linear Regression

Simple linear regression can be extended to include multiple variables (features). This is called multiple linear regression:

$y=\beta_{0}+\beta_{1}x_{1}+...+\beta_{n}x_{n}$

Each $x$ represents a different variable, and each variable has its own coefficient.

For example:

$y=\beta_{0}+\beta_{1}RM+\beta_{2}DIS$


In [29]:
house.columns

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'LSTAT', 'MEDV'],
      dtype='object')

In [30]:
X2 = house[['RM', 'DIS']]
X2 = sm.add_constant(X2) 
y = house['MEDV'] 

model6 = sm.OLS(y, X2).fit() 

predictions6 = model6.predict(X2)

model6.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.496
Model:                            OLS   Adj. R-squared:                  0.494
Method:                 Least Squares   F-statistic:                     247.0
Date:                Sun, 24 Sep 2023   Prob (F-statistic):           1.84e-75
Time:                        10:32:53   Log-Likelihood:                -1667.1
No. Observations:                 506   AIC:                             3340.
Df Residuals:                     503   BIC:                             3353.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -34.6361      2.621    -13.212      0.000     -39.786     -29.486
RM             8.8014      0.424     20.780      0.000       7.969       9.634
DIS            0.4888      0.141      3.459      0.001       0.211       0.767
==============================================================================
Omnibus:                      142.807   Durbin-Watson:                   0.684
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              844.520
Skew:                           1.090   Prob(JB):                    4.12e-184
Kurtosis:                       8.942   Cond. No.                         68.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Interpretation

Interpreting the RM coefficient ($\beta_{1}$) and DIS coefficient ($\beta_{2}$)

A unit increase in RM (average number of rooms in the house) is associated with a $\$$8,800 increase in median house price in that town. 

A unit increase in DIS (distance from house to 5 Boston employment centres) is associated with a $\$$489 increase in median house price in that town. 


## Prediction
Let's say that there was a new house with 6.5 rooms and a distance of 2.3km from the employement centres. What would we predict for the median house price in that town?

$y=-34.6361 + 8.8014 x_{1} + 0.4888 x_{2}$

$y=-34.6361 + 8.8014 \times 6.5 + 0.4888 \times 2.3 = \$ 23,697$



## Prediction using Statsmodels

In [31]:
X_test_point = pd.DataFrame({'const': [1], 'RM': [6.5], 'DIS': [2.3]})[['const', 'RM', 'DIS']]

prediction6 = model6.predict(X_test_point)

print(prediction6)

0    23.697478
dtype: float64


In [32]:
X_test_point

const   RM  DIS
0      1  6.5  2.3